addons/sale/models/sale_order.py

```diff
@@ -305,6 +306,25 @@ def _compute_display_name(self):
                 name = f'{name} - {order.partner_id.name}'
             order.display_name = name
 
+    @api.depends('order_line.product_id')
+    def _compute_has_archived_product_line(self):
+        completed_orders = self.filtered(lambda order: order.state not in ['draft', 'sent'])
+        completed_orders.has_archived_product_line = False
+        non_completed_order = self - completed_orders
+        archived_product_count_per_order = dict(
```

Comment:
Instead of creating a variable and then iterate through non_completed_orders, I think it would be clearer to loop through orders, and in the loop check if the order has archived product and if so, to set the has_archived_product_line to True.

Reply:
I did this because accessing `order_line` in loop means we are doing a query for each order because `order_line` is `one2Many` so if `self` contains 100 orders then it will do 100 queries
Here it'll pre-fetch all SO data in just one query and assign values accordingly
(Didn't really test it)

Reply:
Get it, in this case, I am wondering if it would be possible to somehow set in the archived_product_count_per_order dict in a way that it wouls be [order : boolean]. If it would be possible I think it would be neat solution.

Reply:
Yes sure! done :100: